python/ray/data/block.py

```diff
@@ -274,7 +281,16 @@ def for_block(block: Block) -> "BlockAccessor[T]":
             from ray.data.impl.arrow_block import ArrowBlockAccessor
 
             return ArrowBlockAccessor.from_bytes(block)
+        elif isinstance(block, np.ndarray):
+            from ray.data.impl.arrow_block import ArrowBlockAccessor
+
+            return ArrowBlockAccessor.from_numpy(block)
         elif isinstance(block, list):
+            if block and all(isinstance(item, np.ndarray) for item in block):
+                from ray.data.impl.arrow_block import ArrowBlockAccessor
+
+                return ArrowBlockAccessor.from_numpy(block)
```

Comment:
I don't think this is the right place to do an implicit conversion. You can do that instead in DelegatingBlockBuilder.

Also, this will add performance overhead-- I would push for not doing this in this PR.

Reply:
>I don't think this is the right place to do an implicit conversion. You can do that instead in DelegatingBlockBuilder.

Hmm that may require us to have that conversion logic in a few places, the hope was to centralize it here, similar to the `bytes` case. I'll try to push this up to block building and see if it's reasonable.

Also, are you referring to both the `List[np.ndarray]` and `np.ndarray` cases? Should the `bytes` case technically not be here either, since it's not a true block format?

> Also, this will add performance overhead-- I would push for not doing this in this PR.

This will only have performance overhead for the would-be simple block case, which is already a low performance escape hatch, so I thought that this block traversal would be tenable. But I think that guaranteeing this block traversal only happens in the delegating block builder would probably be better/more isolated.

Reply:
For context, the `bytes` case was a temporary hack due to lack of ArrowTable serialization support proper from Java (x-lang). It should be removed post REP 6 implementation.

Reply:
> Also, are you referring to both the List[np.ndarray] and np.ndarray cases? Should the bytes case technically not be here either, since it's not a true block format?

Yeah-- for clarity, we shouldn't ever have a block of these formats "in the object store".